### PR TITLE
DAOS-4447 control: support rank list with system start/stop

### PR DESCRIPTION
**Pull request by tanabarr**
`dmg system (query|stop|start)` commands should take a `ranks` option
that specifies ranks to operate on.

- extend unit test coverage to verify drpc & grpc calls made on start
- indicate which instances are still running when trying to restart harness
- map ioserver exit error to instance index to aid reporting
- add final instance ready state, distinct from drpc ready 

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/control/common/proto/ctl/addons.go [hunk 1]
```diff
@@ -49,3 +49,45 @@ func (m *SystemQueryReq) GetSystemRanks() []system.Rank {
...
+func (m *SystemStartReq) SetSystemRanks(sysRanks []system.Rank) error {
+	if m == nil {
+		return errors.New("nil request")
+	}
+	return convert.Types(sysRanks, &m.Ranks)
+}
+
+// GetSystemRanks is a convenience method to convert this request's
+// slice of uint32 ranks to a slice of system ranks.
+func (m *SystemStartReq) GetSystemRanks() []system.Rank {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/common/proto/ctl/addons.go:85]: my initial inclination was to do as you suggest with a utility function but I didn't want to go against convention that @mjmac set by specifying the methods on the request pointer receiver. will wait for confirmation before making the change.

File: src/control/cmd/dmg/system_test.go [hunk 2]
```diff
@@ -74,12 +74,60 @@ func TestSystemCommands(t *testing.T) {
...
+			"system stop --rank 0",
+			"",
+			errors.New("unknown flag `rank'"),
+		},
+		{
+			"system stop verbose",
+			"system stop --verbose",
+			"",
+			errors.New("unknown flag `verbose'"),
+		},
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/cmd/dmg/system_test.go:100]: done

File: src/control/common/proto/ctl/addons.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/common/proto/ctl/addons.go:85]: I don't feel strongly either way. It's nice to have them as methods but it does seem to be getting increasingly awkward to have to add them manually. As I noted in another review, I was thinking of these as missing methods that the code generator didn't create. I'll look into whether or not there's any way for us to customize the output, and if not, then perhaps it would be best to consolidate all of these into utility functions.

[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/common/proto/ctl/addons.go:85]: Follow-up: There doesn't seem to be any officially-supported way to customize the code generation, so it's probably best to just extract these into a pair of utility functions as @kjacque suggests.

File: src/control/common/types.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+package common
```
Comment (mjmac) [inline src/control/common/types.go:24]: I was going to object to further inflation of the common package as a grab-bag of random functionality, but it's fine for now. In my control API PR these types are exported as part of the new API.

File: src/control/server/mgmt_pool.go [hunk 4]
```diff
@@ -0,0 +1,340 @@
+//
+// (C) Copyright 2018-2020 Intel Corporation.
```
Comment (mjmac) [inline src/control/server/mgmt_pool.go:2]: If this is a new file, the copyright year should be 2020.

File: src/control/common/proto/ctl/addons.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/common/proto/ctl/addons.go]: https://github.com/daos-stack/daos/pull/2275/ccommits/03c1b159e9c2bad9dfc280dfd6990049e0e1ee7a & https://github.com/daos-stack/daos/pull/2275/commits/9b2a45af12b4d6d5c242fcddffcae66dcd51518c

File: src/control/server/instance.go [hunk 5]
```diff
@@ -90,11 +92,23 @@ func NewIOServerInstance(log logging.Logger,
...
 		scmProvider:       sp,
 		msClient:          msc,
-		instanceReady:     make(chan *srvpb.NotifyReadyReq),
+		drpcReady:         make(chan *srvpb.NotifyReadyReq),
 		storageReady:      make(chan struct{}),
 	}
 }
 
+func (srv *IOServerInstance) setReady() {
+	atomic.StoreUint32(&srv.ready, 1)
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/instance.go:102]: it's not actually in use anywhere in the project but I will replace usages in the server package to make use of it

[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/server/instance.go:102]: It was added in 5ed88b0 to fix a race in storage/scm tests. When you added support for mocking multiple mount points 35189bb, the use of the unprotected map introduced a different race that wasn't caught in review. In fb8385d, I landed a change that added a mutex to protect the map access, and that point there wasn't any point in using `atm.Bool` so it got switched to a regular `bool`.

File: src/control/server/harness.go [hunk 6]
```diff
@@ -414,13 +399,12 @@ func (h *IOServerHarness) RestartInstances() error {
...
 	}
-	if !h.IsStartable() {
-		return errors.New("can't start instances: already running")
-	}
-	if len(h.StartedRanks()) > 0 {
-		return errors.New("can't start instances: already started")
+
+	startedRanks := h.StartedRanks()
+	if len(startedRanks) > 0 {
+		return FaultInstancesNotStopped(startedRanks)
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/harness.go:420]: because of the limitation in the current implementation, we have to restart the harness Start() loop which includes restarting dRPC which will break running process. this is on the list of things to be fixed , but not there yet.

File: src/control/server/instance.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/instance.go]: addressed in https://github.com/daos-stack/daos/pull/2275/commits/4c93f24160a2671deeeaf5af6bf4cb471bd4b140

File: src/control/server/instance.go [hunk 7]
```diff
@@ -491,6 +505,30 @@ func (srv *IOServerInstance) CallDrpc(module, method int32, body proto.Message)
 	return makeDrpcCall(dc, module, method, body)
 }
 
+func (srv *IOServerInstance) SetupWithDrpc(ctx context.Context, ready *srvpb.NotifyReadyReq) error {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/instance.go:510]: done

File: src/control/server/mgmt_pool.go (hunk 4 above)
Comment (tanabarr) [reply src/control/server/mgmt_pool.go]: done

File: src/control/server/ioserver/exec_test.go [hunk 8]
```diff
@@ -124,9 +125,13 @@ func TestRunnerContextExit(t *testing.T) {
 	cancel()
 
 	exitErr := <-errOut
-	if errors.Cause(exitErr) == NormalExit {
+	if errors.Cause(exitErr.Err) == NormalExit {
 		t.Fatal("expected process to not exit normally")
 	}
+	if exitErr.Idx != 9 {
```
Comment (kjacque) [inline src/control/server/ioserver/exec_test.go:131]: nit - should use cfg.Index instead of the literal 9

File: src/control/server/ioserver/exec.go [hunk 9]
```diff
@@ -158,17 +158,9 @@ func (r *Runner) Start(ctx context.Context, errOut chan<- InstanceError) error {
 	return nil
 }
 
-func (r *Runner) setRunning() {
-	atomic.StoreUint32(&r.running, 1)
-}
-
-func (r *Runner) setStopped() {
-	atomic.StoreUint32(&r.running, 0)
-}
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/ioserver/exec.go:154]: seems like making private helper methods that change r.setRunning() from r.running.SetTrue() is overkill, I changed from using atomic to atm packages at @mjmac 's request

File: src/control/system/rank.go [hunk 10]
```diff
@@ -124,3 +126,29 @@ func (r *Rank) RemoveFromList(ranks []Rank) []Rank {
 
 	return rankList
 }
+
+// RanksToUint32 is a convenience method to convert this
+// slice of system ranks to a slice of uint32 ranks.
+func RanksToUint32(ranks []Rank) (uint32Ranks []uint32) {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/system/rank.go:132]: yes agreed, I have a follow-up test coverage PR that is waiting for the client API PR to land that I will include this in

File: src/control/server/ioserver/exec.go [hunk 11]
```diff
@@ -56,6 +56,12 @@ type (
 		running uint32
 		cmd     *exec.Cmd
 	}
+
+	// InstanceError represents error from an instance of a DAOS I/O Server
+	InstanceError struct {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/ioserver/exec.go:61]: we can do but there is currently no need, just need to add the method when it is needed

File: src/control/client/types.go [hunk 12]
```diff
@@ -415,88 +415,3 @@ func (sfr *StorageFormatResult) HasErrors() bool {
 	}
 	return false
 }
-
-// AccessControlList is a structure for the access control list.
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/client/types.go:419]: will do

File: src/control/server/ioserver/exec_test.go (hunk 8 above)
Comment (tanabarr) [reply src/control/server/ioserver/exec_test.go:131]: will do

File: src/control/server/ioserver/exec.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/ioserver/exec.go:154]: leaving for the moment , as we only use them in one place mostly

File: src/control/system/rank.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/system/rank.go:132]: done

File: src/control/server/instance.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/instance.go]: 87dd7ca28ab6bb36229135e61145ad6180ffd756

File: src/control/client/types.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/client/types.go:419]: https://github.com/daos-stack/daos/pull/2365

**Comment by tanabarr** [comment]
gatekeeper please use PR description and title as commit message when landing

File: src/control/server/harness_test.go [hunk 13]
```diff
@@ -260,20 +372,46 @@ func TestServer_HarnessIOServerStart(t *testing.T) {
...
+				0: {fmt.Sprintf("Join %d", system.NilRank)},
+				1: {fmt.Sprintf("Join %d", system.NilRank)},
+			},
+			expRanks: map[uint32]system.Rank{
+				0: system.Rank(0),
+				1: system.Rank(1),
+			},
+			expIoErrs: map[uint32]error{
+				0: errors.New("oops"),
+				1: errors.New("oo"),
```
Comment (kjacque) [inline src/control/server/harness_test.go:406]: Should be "oops"?

Comment (tanabarr) [reply src/control/server/harness_test.go:406]: yes it should . but this actually passes because CmpErr only checks that the text is a substring of the error, not exact match.

File: src/control/server/harness_test.go [hunk 14]
```diff
@@ -302,58 +444,178 @@ func TestServer_HarnessIOServerStart(t *testing.T) {
...
+					}
+					<-time.After(testShortTimeout)
+				}
+			}()
+
+			select {
+			case <-waitDrpcReady:
+			case <-ctx.Done():
+				if tc.expStartErr != context.DeadlineExceeded {
+					CmpErr(t, tc.expStartErr, gotErr)
```
Comment (kjacque) [inline src/control/server/harness_test.go:532]: The race condition you're seeing has to do with this line. I think you need to wait on the "done" channel in this case, before looking at gotErr.

**Comment by tanabarr** [comment]
for some reason . clicking on the "continuous-integration/jenkins/pr-head" link above does not take you to https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-2275/22/pipeline as it should

File: src/control/server/harness_test.go [hunk 15]
```diff
@@ -235,23 +240,132 @@ func TestServer_HarnessGetMSLeaderInstance(t *testing.T) {
...
+		instanceUuids    map[int]string // UUIDs for each instance.Index()
+		expStartErr      error
+		expStartCount    int
+		expDrpcCalls     map[uint32][]int32     // method ids called for each instance.Index()
+		expGrpcCalls     map[uint32][]string    // string repr of call for each instance.Index()
+		expRanks         map[uint32]system.Rank // ranks to have been set during Start()
+		expMembers       system.Members         // members to have been registered during Stop()
+		expIoErrs        map[uint32]error       // errors expected from instances
 	}{
 		"normal startup/shutdown": {
```
Comment (kjacque) [inline src/control/server/harness_test.go:271]: Looking a little more closely at these tests, still a little hard to follow. I guess it's pretty weird to me that some cases are normal startup/shutdown but still return errors. Would be nice to add some commentary about why we're triggering errors in a normal case - Are we just using the context timeout to end the test case?

Comment (tanabarr) [reply src/control/server/harness_test.go:271]: yes, because they wouldn't normally end in normal circumstances

Comment (kjacque) [reply src/control/server/harness_test.go:271]: Ah, got it. Thanks for the details, Tom.

**Comment by tanabarr** [comment]
Jenkins CI runs for this PR's changes (2383 is a clone of 2275) showing only known intermittent failures:
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-2383/1/tests/
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-2275/22/tests/
